File: core/src/net/bossdragon/util/operations/Operations.java
package net.bossdragon.util.operations;

import com.badlogic.gdx.math.Interpolation;
import com.badlogic.gdx.math.Vector2;
import static se.feomedia.orion.OperationFactory.*;

/**
 *
 */
public class Operations {
    public static MoveByOp moveBy(Vector2 vec, float duration, Interpolation interpolation) {
        MoveByOp op = operation(MoveByOp.class);
        configure(op, duration, interpolation);
        op.dst.set(vec);

        return op;
    }

    public static MoveByOp moveBy(Vector2 vec, float duration) {
        return moveBy(vec, duration, Interpolation.linear);
    }

    public static MoveByOp moveBy(Vector2 vec) {
        return moveBy(vec, 0f);
    }

    public static MoveToOp moveTo(Vector2 vec, float duration, Interpolation interpolation) {
        MoveToOp op = operation(MoveToOp.class);
        configure(op, duration, interpolation);
        op.dst.set(vec);

        return op;
    }

    public static MoveToOp moveTo(Vector2 vec, float duration) {
        return moveTo(vec, duration, Interpolation.linear);
    }

    public static MoveToOp moveTo(Vector2 vec) {
        return moveTo(vec, 0f, Interpolation.linear);
    }

    public static RotateByOp rotateBy(float angle, float duration, Interpolation interpolation) {
        RotateByOp op = operation(RotateByOp.class);
        configure(op, duration, interpolation);
        op.dst = angle;

        return op;
    }

    public static RotateByOp rotateBy(float angle, float duration) {
        return rotateBy(angle, duration, Interpolation.linear);
    }

    public static RotateByOp rotateBy(float angle) {
        return rotateBy(angle, 0f, Interpolation.linear);
    }

    public static RotateToOp rotateTo(float angle, float duration, Interpolation interpolation) {
        RotateToOp op = operation(RotateToOp.class);
        configure(op, duration, interpolation);
        op.dst = angle;

        return op;
    }

    public static RotateToOp rotateTo(float angle, float duration) {
        return rotateTo(angle, duration, Interpolation.linear);
    }

    public static RotateToOp rotateTo(float angle) {
        return rotateTo(angle, 0f, Interpolation.linear);
    }

    public static DisplaceByOp displaceBy(float x, float y, float duration, Interpolation interpolation) {
        DisplaceByOp op = operation(DisplaceByOp.class);
        configure(op, duration, interpolation);
        op.dst.set(x, y);

        return op;
    }

    public static DisplaceByOp displaceBy(float x, float y) {
        return displaceBy(x, y, 0f, Interpolation.linear);
    }

    public static DisplaceByOp displaceBy(Vector2 translation) {
        return displaceBy(translation.x, translation.y);
    }

    public static DisplaceByOp displaceBy(Vector2 translation, float duration) {
        return displaceBy(translation.x, translation.y, duration, Interpolation.linear);
    }

    public static DisplaceByOp displaceBy(Vector2 translation, float duration, Interpolation interpolation) {
        return displaceBy(translation.x, translation.y, duration, interpolation);
    }


    public static DisplaceToOp displaceTo(Vector2 dest, float duration, Interpolation interpolation) {
        DisplaceToOp op = operation(DisplaceToOp.class);
        configure(op, duration, interpolation);
        op.dst.set(dest);

        return op;
    }

    public static DisplaceToOp displaceTo(Vector2 dest, float duration) {
        return displaceTo(dest, duration, Interpolation.linear);
    }

    public static OpacityToOp opacityTo(float targetOpacity, float duration, Interpolation interpolation) {
        OpacityToOp op = operation(OpacityToOp.class);
        configure(op, duration, interpolation);
        op.dst = targetOpacity;

        return op;
    }

    public static OpacityToOp opacityTo(float targetOpacity, float duration) {
        return opacityTo(targetOpacity, duration, Interpolation.linear);
    }

    public static TemporalOp temporal(float duration, TemporalOp.SimpleExecutor executor) {
        TemporalOp op = operation(TemporalOp.class);
        op.duration = duration;
        op.executor = executor;

        return op;
    }

    public static TemporalOp temporal(float duration, Interpolation interpolation, TemporalOp.SimpleExecutor executor) {
        final TemporalOp op = temporal(duration, executor);
        op.interpolation = interpolation;

        return op;
    }

    public static TemporalDeltaOp deltaTemporal(float duration, TemporalDeltaOp.DeltaExecutor executor) {
        TemporalDeltaOp op = operation(TemporalDeltaOp.class);
        op.duration = duration;
        op.executor = executor;

        return op;
    }

    public static TemporalDeltaOp deltaOperation(float duration, Interpolation interpolation, TemporalDeltaOp.DeltaExecutor executor) {
        TemporalDeltaOp op = deltaTemporal(duration, executor);
        op.interpolation = interpolation;

        return op;
    }

    public static CallOp call(Runnable callback) {
        CallOp op = operation(CallOp.class);
        op.callback = callback;

        return op;
    }
}

File: core/src/net/bossdragon/util/operations/Operations.kt
/**
 * Project-specific operations to be used with artemis-odb-orion
 */
package net.bossdragon.util.operations.funcs

import com.badlogic.gdx.math.Interpolation
import com.badlogic.gdx.math.Vector2
import net.bossdragon.util.operations.*
import se.feomedia.orion.OperationFactory.configure
import se.feomedia.orion.OperationFactory.operation


fun moveBy(vec: Vector2, duration: Float, interpolation: Interpolation = Interpolation.linear): MoveByOperation {
    val op = operation(MoveByOperation::class.java)
    configure(op, duration, interpolation)
    op.dest.set(vec)

    return op
}

fun moveBy(vec: Vector2): MoveByOperation {
    return moveBy(vec, 0f, Interpolation.linear)
}

fun moveTo(destination: Vector2, duration: Float, interpolation: Interpolation = Interpolation.linear): MoveToOperation {
    val op = operation(MoveToOperation::class.java)
    configure(op, duration, interpolation)
    op.dest.set(destination)

    return op
}

fun moveTo(destination: Vector2): MoveToOperation {
    return moveTo(destination, 0f, Interpolation.linear)
}

fun rotateBy(angle: Float, duration: Float, interpolation: Interpolation = Interpolation.linear): RotateByOperation {
    val op = operation(RotateByOperation::class.java)
    configure(op, duration, interpolation)
    op.dest = angle

    return op
}

fun rotateBy(angle: Float): RotateByOperation {
    return rotateBy(angle, 0f, Interpolation.linear)
}

fun rotateTo(angle: Float, duration: Float, interpolation: Interpolation = Interpolation.linear): RotateToOperation {
    val op = operation(RotateToOperation::class.java)
    configure(op, duration, interpolation)
    op.dest = angle

    return op
}

fun rotateTo(angle: Float): RotateToOperation {
    return rotateTo(angle, 0f, Interpolation.linear)
}

fun displaceBy(translation: Vector2, duration: Float, interpolation: Interpolation = Interpolation.linear): DisplaceByOperation {
    val op = operation(DisplaceByOperation::class.java)
    configure(op, duration, interpolation)
    op.dest.set(translation)

    return op
}

fun displaceBy(x: Float, y: Float, duration: Float, interpolation: Interpolation = Interpolation.linear): DisplaceByOperation {
    val op = operation(DisplaceByOperation::class.java)
    configure(op, duration, interpolation)
    op.dest.set(x, y)

    return op
}

fun displaceBy(translation: Vector2): DisplaceByOperation {
    return displaceBy(translation, 0f, Interpolation.linear)
}

fun displaceBy(x: Float, y: Float): DisplaceByOperation {
    return displaceBy(x, y, 0f, Interpolation.linear)
}

fun displaceTo(destination: Vector2, duration: Float, interpolation: Interpolation = Interpolation.linear): DisplaceToOperation {
    val op = operation(DisplaceToOperation::class.java)
    configure(op, duration, interpolation)
    op.dest.set(destination)

    return op
}

fun opacityTo(targetOpacity: Float, duration: Float, interpolation: Interpolation = Interpolation.linear): OpacityToOperation {
    val op = operation(OpacityToOperation::class.java)
    configure(op, duration, interpolation)
    op.dst = targetOpacity

    return op
}

fun displaceTo(translation: Vector2): DisplaceToOperation {
    return displaceTo(translation, 0f, Interpolation.linear)
}

fun temporal(duration: Float, executor: TemporalOperation.SimpleExecutor): TemporalOperation {
    val action = operation(TemporalOperation::class.java)
    action.duration = duration
    action.executor = executor

    return action
}

fun temporal(duration: Float, interpolation: Interpolation, executor: TemporalOperation.SimpleExecutor): TemporalOperation {
    val action = temporal(duration, executor)
    action.interpolation = interpolation

    return action
}

fun deltaTemporal(duration: Float, executor: TemporalDeltaOperation.DeltaExecutor): TemporalDeltaOperation {
    val action = operation(TemporalDeltaOperation::class.java)
    action.duration = duration
    action.executor = executor

    return action
}

fun deltaOperation(duration: Float, interpolation: Interpolation, executor: TemporalDeltaOperation.DeltaExecutor): TemporalDeltaOperation {
    val action = deltaTemporal(duration, executor)
    action.interpolation = interpolation

    return action
}

fun call(callback: Runnable): CallOperation {
    val action = operation(CallOperation::class.java)
    action.callback = callback

    return action
}
